src/Microsoft.AspNet.Security.OAuth/OAuthBearerAuthenticationHandler.cs

```diff
@@ -24,87 +35,167 @@ protected override AuthenticationTicket AuthenticateCore()
-                    Context,
-                    Options.AccessTokenFormat,
-                    requestTokenContext.Token);
-
-                await Options.AccessTokenProvider.ReceiveAsync(tokenReceiveContext);
-                if (tokenReceiveContext.Ticket == null)
+				// notify user token was received
+				var securityTokenReceivedNotification = 
+					new SecurityTokenReceivedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>(Context, Options)
+					{
+						ProtocolMessage = requestTokenContext,
+					};
+
+				await Options.Notifications.SecurityTokenReceived(securityTokenReceivedNotification);
+				if (securityTokenReceivedNotification.HandledResponse)
+				{
+					return securityTokenReceivedNotification.AuthenticationTicket;
+				}
+
+				if (securityTokenReceivedNotification.Skipped)
+				{
+					return null;
+				}
+
+				if (_configuration == null)
                 {
-                    tokenReceiveContext.DeserializeTicket(tokenReceiveContext.Token);
+                    _configuration = await Options.ConfigurationManager.GetConfigurationAsync(Context.RequestAborted);
                 }
 
-                AuthenticationTicket ticket = tokenReceiveContext.Ticket;
-                if (ticket == null)
-                {
-                    _logger.WriteWarning("invalid bearer token received");
-                    return null;
-                }
+				var validationParameters = Options.TokenValidationParameters.Clone();
+				if (validationParameters.ValidIssuer == null && !string.IsNullOrWhiteSpace(_configuration.Issuer))
+				{
+					validationParameters.ValidIssuer = _configuration.Issuer;
+				}
+				else
+				{
+					IEnumerable<string> issuers = new[] { _configuration.Issuer };
+					validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? issuers : validationParameters.ValidIssuers.Concat(issuers));
+				}
+
+				validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? _configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(_configuration.SigningKeys));
+				SecurityToken validatedToken;
+				foreach (var validator in Options.SecurityTokenValidators)
+				{
+					if (validator.CanReadToken(requestTokenContext.Token))
+					{
+
+						ClaimsPrincipal principal = Options.SecurityTokenValidators.First().ValidateToken(requestTokenContext.Token, validationParameters, out validatedToken);
+						ClaimsIdentity claimsIdentity = principal.Identity as ClaimsIdentity;
+						AuthenticationTicket ticket = new AuthenticationTicket(claimsIdentity, new AuthenticationProperties());
+						var securityTokenValidatedNotification = new SecurityTokenValidatedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>(Context, Options)
+						{
+							ProtocolMessage = requestTokenContext,
+							AuthenticationTicket = ticket
+						};
+
+						if (securityTokenReceivedNotification.HandledResponse)
```

Comment:
Do we need to raise `SecurityTokenValidated` here and check if handled response on `securityTokenValidatedNotification`?